standards/core/rln-contract.md

```diff
@@ -0,0 +1,204 @@
+1. A user registers a membership:
+    1. If there are expired unclaimed slots in the tree, use the slot with the earliest expiration date past its grace period, otherwise create a new slot;
+    2. if the membership re-uses a previously used slot, refund the previous user’s deposit.
+2. The user sends messages while the membership is active:
+    1. If the user exceeds the rate limit, their over-limit messages are not propagated in that epoch; the deposit is not slashed.
+3. After the membership expires, the user can do one of the following:
+    1. Do nothing. The deposit remains in the contract, and the user retains the ability to send messages (respecting the per-epoch rate limit);
+    2. Withdraw the deposit. The user receives the full deposit back, and loses the ability to send messages;
+    3. Extend the membership. The user sends a transaction to re-enable their membership for another expiration term. The user only covers gas costs and does not have to provide additional funds. The original deposit remains in the contract. The membership is prolonged for another expiration term under the same conditions.
+
+The user can extend the membership both during and after its grace period.
+By not extending the membership within its grace period, the user assumes the risk of the membership being taken over at any moment.
+
+A user can hold multiple memberships.
+Memberships are not transferable.
+There is no limitation on the number of membership per user (apart from the global limits, see parameter table).
+
+## Contract governance and mutability
+
+The contract governance is follows:
+
+1. the first deployment of the contract allows the `Owner` to modify certain parameters (TBD) under certain constraints (TBD);
+2. at some point, the `Onwer` would renounce their privileges, and the contract will become immutable;
+3. further upgrades, if necessary, can be done by deploying a new contract and moving the membership set.
+
+## Parameters
+
+We make the following design choices:
+- there are `3` membership tiers: low-, mid-, and high-tier;
+- prices are quoted in `USD`;
+- payment is accepted in `DAI`;
+- the price for a given tier is linearly proportional to its rate limit.
+
+For the initial deployment, we suggest the following values.
+
+| Parameter                                                 | Symbol     | Value   | Units                             | Modifiable? | Comment                        |
+| --------------------------------------------------------- | ---------- | ------- | --------------------------------- | ----------- | ------------------------------ |
+| Epoch length                                              | `epoch`    | `10`    | minutes                           | TBD         |                                |
+| Maximum total rate limit of concurrent active memberships | `R_{max}`  | TBD     | messages per `epoch`              | Yes         | Can be replaced with `N_{max}` |
```

Comment:
Is that a contract parameter? or a calculation deduced from other parameters.

Actually, may be interesting to provide 2 values deduced from parameters:
- sustainable bandwidth usage: the bandwidth usage over several epoch if all members uses the max of their limit evenly distributed on epoch, probably multiplied by recommended gossipsub parameters
- max spike bandwidth usage: bandwidth usage if all members max out their rate limit at the same time, the fastest way possible.

Reply:
This is supposed to be a contract parameter, perhaps I should add a paragraph to make it clearer. My idea was that the contract should limit the total concurrent RLN bandwidth. If this is something we want to do, I see two ways to encode this restriction:

- limit the total number of concurrently active memberships;
- limit the total rate limit of those memberships.

The question is which of this parameters, or both, we should use.